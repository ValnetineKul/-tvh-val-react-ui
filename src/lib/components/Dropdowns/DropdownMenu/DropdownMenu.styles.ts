import { makeStyles } from '../../../../themes/core';

type Props = {
  maxListHeight: number;
};

const useStyles = makeStyles<Props>({ name: 'DropdownMenu' })((theme, props) => ({
  popperRoot: {
    zIndex: 1291,
  },
  mainSurface: {
    marginTop: theme.layoutSpacing['spacing/200'],
    marginBottom: theme.layoutSpacing['spacing/200'],
  },
  menu: {
    borderRadius: theme.borderRadius['borderRadius/none'],
    boxShadow: theme.shadow['shadow/300'],
    display: 'block',
    overflowY: 'auto',
    maxHeight: 320,
    maxWidth: 320,
    minWidth: 120,
  },
  dynamicList: {
    maxHeight: props.maxListHeight,
  },
  menuList: {
    padding: `${theme.layoutSpacing['spacing/300']} ${theme.layoutSpacing['spacing/000']}`,
  },
  subMenu: {
    zIndex: 1300,
    marginTop: theme.layoutSpacing['spacing/000'],
    '&[x-placement*="left"]': {
      clipPath: 'inset(-24px -1px -24px -24px)',
    },
    '&[x-placement*="right"]': {
      clipPath: 'inset(-24px -24px -24px -1px)',
    },
  },
  mobileSubMenu: {
    display: 'flex',
    alignItems: 'center',
    width: '100%',
  },
  fullSpinner: {
    display: 'block',
  },
  emptySpinnerContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },

  mobileEmptySpinnerContainer: {
    height: 72,
  },
  modalAutoHeight: {
    height: 'auto',
    maxHeight: '95vh',
    top: 'auto',
  },
  modalTitle: {
    marginLeft: 'auto',
    marginRight: 'auto',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    whiteSpace: 'nowrap',
  },
  title: {
    padding: `${theme.layoutSpacing['spacing/200']} ${theme.layoutSpacing['spacing/400']}`,
  },
  modalContainer: {
    height: 'auto',
  },
  modalScrollList: {
    overflowY: 'auto',
  },
  mobileMenuList: {
    paddingTop: theme.layoutSpacing['spacing/000'],
    paddingBottom: theme.layoutSpacing['spacing/000'],
    height: '100%',
    overflowY: 'auto',
  },
  modalLabel: {
    display: 'block',
    width: '100%',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    textAlign: 'center',
    marginRight: theme.layoutSpacing['spacing/300'],
    marginLeft: theme.layoutSpacing['spacing/300'],
    '&::first-letter': {
      textTransform: 'capitalize',
    },
  },
  hiddenContentFromScreenReader: {
    visibility: 'hidden',
  },
}));

export default useStyles;
