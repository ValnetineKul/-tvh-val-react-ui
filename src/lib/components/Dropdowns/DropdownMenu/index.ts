import DropdownMenu from './DropdownMenu';

export type { DropdownMenuProps } from './DropdownMenu';
export default DropdownMenu;
