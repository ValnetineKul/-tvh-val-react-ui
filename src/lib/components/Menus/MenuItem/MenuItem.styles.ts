import type { Surface } from '../../../../themes/core';
import { makeStyles } from '../../../../themes/core';

type Props = {
  surface: Surface;
};

const useStyles = makeStyles<
  Props,
  'wrapper' | 'selected' | 'disabled' | 'startIcon' | 'endIcon' | 'label' | 'subLabel' | 'tag' | 'inlineMessage'
>({ name: 'MenuItem' })((theme, { surface }, classes) => ({
  root: {
    padding: theme.layoutSpacing['spacing/000'],
    minHeight: 'auto',
    '& label': {
      width: '100%',
      padding: `${theme.layoutSpacing['spacing/350']} ${theme.layoutSpacing['spacing/400']}`,
    },
    [`&.${classes.disabled}`]: {
      opacity: 1,
      [`& .${classes.subLabel}, & .${classes.inlineMessage} svg, & .${classes.startIcon}, & .${classes.endIcon}`]: {
        color: theme.color[`text/onSurface${surface}/disabled` as const],
      },
      [`& .${classes.tag}`]: {
        backgroundColor: theme.color['bg/disabled'],
        color: theme.color['text/onSurface100/disabled'],
      },
    },
    [`&.${classes.selected}`]: {
      backgroundColor: 'transparent',
    },
    [`&.${classes.selected}:hover, &:not(:disabled):hover`]: {
      backgroundColor: theme.color[`bg/surface${surface}/hover` as const],
    },
    [`&.${classes.selected}:active, &:not(:disabled):active`]: {
      backgroundColor: theme.color[`bg/surface${surface}/active` as const],
    },
  },
  wrapper: {
    padding: `${theme.layoutSpacing['spacing/350']} ${theme.layoutSpacing['spacing/400']}`,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: '100%',
    minWidth: 120,
    minHeight: theme.layoutSpacing['spacing/550'],
    '&:not(:disabled):hover': {
      backgroundColor: 'transparent',
    },
  },
  labelWrapper: {
    flex: 1,
    overflow: 'hidden',
    textAlign: 'left',
  },
  startIcon: {
    marginRight: theme.layoutSpacing['spacing/300'],
    color: theme.color[`icon/onSurface${surface}/default` as const],
  },
  endIcon: {
    marginLeft: theme.layoutSpacing['spacing/300'],
    color: theme.color[`icon/onSurface${surface}/default` as const],
  },
  label: {
    display: 'block',
    width: '100%',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    '&::first-letter': {
      textTransform: 'capitalize',
    },
  },
  tag: {
    marginLeft: theme.layoutSpacing['spacing/300'],
  },
  checkboxWithTagWrapper: {
    display: 'flex',
    justifyContent: 'space-between',
  },
  subLabel: {},
  selected: {},
  disabled: {},
  inlineMessage: {},
  focusVisible: {
    [`& .${classes.wrapper}`]: {
      backgroundColor: theme.color[`bg/surface${surface}/hover` as const],
    },
  },
}));

export default useStyles;
