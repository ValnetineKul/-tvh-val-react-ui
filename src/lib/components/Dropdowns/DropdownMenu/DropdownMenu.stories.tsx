import { action } from '@storybook/addon-actions';
import React from 'react';
import { createMeta } from '../../../story-utils';
import type { StoryTemplate } from '../../../story-utils';
import { AngleRight } from '../../Icon/icons/functional';
import Icon from '../../Icon';
import MenuItem from '../../Menus/MenuItem';
import DropdownMenu from './DropdownMenu';

type Props = React.ComponentProps<typeof DropdownMenu>;

const Template = ((args: Props) => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);

  return (
    <>
      <div ref={setAnchorEl} />
      <DropdownMenu {...args} anchor={anchorEl} onClose={() => {}} />
    </>
  );
}) as StoryTemplate<Props>;

export const Primitive = Template.bind({});
Primitive.args = {
  mobileMenuTitle: 'Dropdown header',
  menuItems: [
    <MenuItem label="very long menu item name which should be truncated" onClick={action('onItemClick')} />,
    <MenuItem label="item_2" onClick={action('onItemClick')} />,
    <MenuItem
      label="item_3"
      endIcon={<Icon icon={AngleRight} />}
      onClick={action('onItemClick')}
      subMenu={[
        <MenuItem label="sub_item_1" onClick={action('onItemClick')} />,
        <MenuItem label="sub_item_2" onClick={action('onItemClick')} />,
      ]}
    />,
    <MenuItem label="item_4" onClick={action('onItemClick')} />,
    <MenuItem label="item_5" onClick={action('onItemClick')} />,
    <MenuItem label="item_6" onClick={action('onItemClick')} />,
    <MenuItem label="item_7" onClick={action('onItemClick')} />,
    <MenuItem label="item_8" onClick={action('onItemClick')} />,
    <MenuItem label="item_9" onClick={action('onItemClick')} />,
    <MenuItem label="item_10" onClick={action('onItemClick')} />,
    <MenuItem label="item_11" onClick={action('onItemClick')} />,
    <MenuItem label="item_12" onClick={action('onItemClick')} />,
    <MenuItem label="item_13" onClick={action('onItemClick')} />,
    <MenuItem label="item_14" onClick={action('onItemClick')} />,
    <MenuItem label="item_15" onClick={action('onItemClick')} />,
  ],
  position: 'bottom-start',
  onClose: () => {},
  isLoading: false,
};

Primitive.storyName = 'Menu';

export const PrimitiveLoadingMenu = Template.bind({});
PrimitiveLoadingMenu.args = {
  mobileMenuTitle: 'Dropdown header',
  menuItems: [],
  position: 'bottom-start',
  onClose: () => {},
  isLoading: true,
};

PrimitiveLoadingMenu.storyName = 'LoadingMenu';

export default createMeta({
  component: DropdownMenu,
  title: 'Components/Dropdowns/Menu',
  argTypes: {
    menuItems: {
      control: {
        disable: true,
      },
      table: {
        type: {
          summary: 'An array of MenuItem components',
        },
      },
    },
    anchor: {
      control: {
        disable: true,
      },
      table: {
        type: {
          summary: 'HTMLElement, used to set the position of the Dropdown',
        },
      },
    },
  },
});
