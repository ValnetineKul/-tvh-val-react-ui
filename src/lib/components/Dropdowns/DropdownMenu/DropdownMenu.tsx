import ClickAwayListener from '@mui/material/ClickAwayListener';
import Popper from '@mui/material/Popper';

import type { FC } from 'react';
import React, { useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import useScreenSize from '../../../hooks/useScreenSize';
import Surface from '../../Surfaces/Surface';
import MenuItem from '../../Menus/MenuItem';
import MenuList from '../../Menus/MenuList';
import Spinner from '../../ProgressIndicators/Spinner';
import SpinnerWithBackdrop from '../../ProgressIndicators/Spinner/SpinnerWithBackdrop';
import Modal from '../../../patterns/Dialogs/Modal';
import Typography from '../../Typography';
import { AngleLeft, Times } from '../../Icon/icons/functional';
import Icon from '../../Icon';
import IconButton from '../../Buttons/IconButton';
import { useComponentCountController } from '../../../hooks/useComponentCounterHooks/useComponentCounterHooks';
import type { MenuItemProps } from '../../Menus/MenuItem/MenuItem.types';
import useDropdownHeightCalculator from '../DropdownList/useDropdownHeightCalculator';
import useStyles from './DropdownMenu.styles';
import {
  DEFAULT_VISIBLE_ITEMS_COUNT,
  LARGE_ITEM_HEIGHT,
  DEFAULT_ITEM_HEIGHT,
} from '../DropdownList/DropdownList.constants';

const LIST_CONTAINER_OFFSET = 40;

function getMenuItemHeight<T extends { props: MenuItemProps }[]>(items: T) {
  const itemData = React.Children.toArray(items) as T;

  return itemData[0].props.subLabel ? LARGE_ITEM_HEIGHT : DEFAULT_ITEM_HEIGHT;
}

export interface DropdownMenuProps {
  anchor?: HTMLElement | null;
  menuItems: (React.ReactElement<React.ComponentProps<typeof MenuItem>> | null)[];
  visibleItemsCount?: number;
  position?: 'bottom-end' | 'bottom-start' | 'bottom' | 'top-end' | 'top-start' | 'top';
  onClose: () => void;
  isLoading?: boolean;
  isSubMenuLoading?: boolean;
  mobileMenuTitle?: string;
  menuListClassName?: string;
  modalClassName?: string;
}

const DropdownMenu: FC<DropdownMenuProps> = ({
  anchor,
  menuItems,
  visibleItemsCount: visibleItemsCountProp,
  position = 'bottom-start',
  onClose,
  isLoading = false,
  isSubMenuLoading = false,
  mobileMenuTitle = 'Dropdown header',
  menuListClassName,
  modalClassName,
}) => {
  const visibleItemsCount = visibleItemsCountProp || DEFAULT_VISIBLE_ITEMS_COUNT;
  const {
    anchorPosition,
    listHeight: estimatedListHeight,
    maxListHeight,
    setListContainer,
  } = useDropdownHeightCalculator(
    anchor,
    null,
    visibleItemsCount,
    menuItems.length > 0 ? getMenuItemHeight(menuItems) : 0,
    LIST_CONTAINER_OFFSET
  );
  const { classes, cx } = useStyles({
    maxListHeight: !!visibleItemsCountProp && estimatedListHeight < maxListHeight ? estimatedListHeight : maxListHeight,
  });
  const { isTabletUp } = useScreenSize('Tablet');

  const [isSubMenuOpen, setIsSubMenuOpen] = useState(false);
  const [subMenuAnchor, setSubMenuAnchor] = useState<null | HTMLElement>(null);

  const shouldShowMaxListItems = isTabletUp;

  const toggleSubMenu = (event: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>) => {
    setIsSubMenuOpen(!isSubMenuOpen);
    setSubMenuAnchor(event.currentTarget);
  };

  const handleMouseOver = (event: React.MouseEvent<HTMLLIElement>) => {
    setIsSubMenuOpen(true);
    setSubMenuAnchor(event.currentTarget);
  };

  const handleCloseList = (event: MouseEvent | TouchEvent) => {
    if (anchor && anchor.contains(event.target as HTMLElement)) {
      return;
    }
    onClose();
  };

  const renderModalTitle = () => {
    return (
      <>
        {typeof mobileMenuTitle === 'string' ? (
          <Typography variant="h5" className={classes.modalTitle} component="span">
            {mobileMenuTitle}
          </Typography>
        ) : (
          mobileMenuTitle
        )}
      </>
    );
  };

  const renderList = () => {
    if (isLoading && menuItems.length === 0) {
      return (
        <li
          className={
            isTabletUp
              ? classes.emptySpinnerContainer
              : cx(classes.mobileEmptySpinnerContainer, classes.emptySpinnerContainer)
          }
        >
          <Spinner />
        </li>
      );
    }

    return menuItems.map((item, key) => {
      if (!item) {
        return null;
      }

      const { label, disabled: isDisabled, subMenu } = item.props;

      if (subMenu) {
        const renderSubMenu = (sublevelLabel: string | JSX.Element) => {
          const renderSubMenuList = () => {
            return (
              <MenuList className={isTabletUp ? classes.menuList : classes.mobileMenuList}>
                {subMenu?.map((subMenuItem, index) => React.cloneElement(subMenuItem, { key: index }))}
              </MenuList>
            );
          };

          const renderSubModalTitle = () => {
            return (
              <div className={classes.mobileSubMenu}>
                <IconButton
                  aria-label="back"
                  size="md"
                  onClick={() => setIsSubMenuOpen(false)}
                  icon={<Icon icon={AngleLeft} />}
                />
                {typeof sublevelLabel === 'string' ? (
                  <Typography variant="h5" component="span" className={classes.modalLabel}>
                    {sublevelLabel}
                  </Typography>
                ) : (
                  sublevelLabel
                )}
                <IconButton
                  aria-label="close"
                  size="md"
                  onClick={() => setIsSubMenuOpen(false)}
                  icon={<Icon icon={Times} />}
                />
              </div>
            );
          };

          if (isTabletUp) {
            return (
              <Popper
                open={!isDisabled && !!anchor && isSubMenuOpen}
                anchorEl={subMenuAnchor}
                className={cx(classes.menu, classes.subMenu)}
                placement="left-start"
              >
                <Surface color="100">{renderSubMenuList()}</Surface>
              </Popper>
            );
          }
          return (
            <Modal
              open={isSubMenuOpen}
              id={typeof sublevelLabel === 'string' ? sublevelLabel : uuidv4()}
              title={renderSubModalTitle()}
              rootClassName={classes.modalAutoHeight}
              containerClassName={classes.modalContainer}
              titleClassName={classes.title}
            >
              <SpinnerWithBackdrop
                className={cx(classes.fullSpinner, classes.modalScrollList)}
                isLoading={isSubMenuLoading && menuItems.length > 0}
              >
                {renderSubMenuList()}
              </SpinnerWithBackdrop>
            </Modal>
          );
        };

        // exclude checkbox specific props
        const {
          onCheckboxChange,
          checkboxProps,
          checkbox: hasCheckbox,
          indeterminate: isIndeterminate,
          ...props
        } = item.props;

        return (
          // workaround because of Nested Menu component is planned to be developed: https://material-ui.com/discover-more/roadmap/#main-content
          // keyboard navigation for submenu does not work
          // TODO: improve subMenu (desktop and mobile version) to make it semantical and accessible
          /* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
          <li
            key={key}
            onMouseEnter={isTabletUp ? handleMouseOver : undefined}
            onClick={isTabletUp ? handleMouseOver : undefined}
            onKeyUp={isTabletUp ? () => handleMouseOver : undefined}
            onMouseLeave={isTabletUp ? () => setIsSubMenuOpen(false) : undefined}
          >
            <MenuItem {...props} label={label} elementType="div" disabled={isDisabled} onClick={toggleSubMenu} />
            {renderSubMenu(label)}
          </li>
          /* eslint-enable jsx-a11y/no-noninteractive-element-interactions */
        );
      }

      return React.cloneElement(item, { key });
    });
  };

  const renderSpinnerWithBackdrop = () => {
    return (
      <SpinnerWithBackdrop
        ref={setListContainer}
        className={cx(classes.fullSpinner, { [classes.modalScrollList]: !isTabletUp })}
        isLoading={isLoading && menuItems.length > 0}
      >
        <MenuList
          className={
            isTabletUp
              ? cx(classes.menu, classes.menuList, { [classes.dynamicList]: shouldShowMaxListItems }, menuListClassName)
              : classes.mobileMenuList
          }
        >
          {renderList()}
        </MenuList>
      </SpinnerWithBackdrop>
    );
  };

  useComponentCountController('DropdownMenu:mobile', !isTabletUp && !!anchor);

  if (isTabletUp) {
    return (
      <ClickAwayListener onClickAway={handleCloseList}>
        <Popper
          className={classes.popperRoot}
          open={!!anchor}
          anchorEl={anchor}
          placement={position || (anchorPosition === 'top' ? 'bottom-end' : 'top-end')}
        >
          <Surface color="100" className={classes.mainSurface}>
            {renderSpinnerWithBackdrop()}
          </Surface>
        </Popper>
      </ClickAwayListener>
    );
  }

  return (
    <Modal
      open={!!anchor}
      id={typeof mobileMenuTitle === 'string' ? mobileMenuTitle : uuidv4()}
      title={renderModalTitle()}
      onClose={onClose}
      rootClassName={cx(
        classes.modalAutoHeight,
        { [classes.hiddenContentFromScreenReader]: isSubMenuOpen },
        modalClassName
      )}
      containerClassName={classes.modalContainer}
      titleClassName={classes.title}
    >
      {renderSpinnerWithBackdrop()}
    </Modal>
  );
};

export default DropdownMenu;
