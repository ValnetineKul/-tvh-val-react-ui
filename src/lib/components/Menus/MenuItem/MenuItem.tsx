import React, { forwardRef, isValidElement } from 'react';
import { default as MuiMenuItem } from '@mui/material/MenuItem';
import TruncatedText from '../../TruncatedText/TruncatedText';
import ButtonBase from '../../Surfaces/ButtonBase';
import Checkbox from '../../Checkbox';
import useStyles from './MenuItem.styles';
import { useButtonBase } from '../../ButtonBase';
import type { MenuItemProps } from './MenuItem.types';

const MenuItem = forwardRef<HTMLLIElement, MenuItemProps>(
  (
    {
      label,
      selected = false,
      disabled = false,
      elementType = 'li',
      labelClassName,
      className,
      focusVisibleClassName,
      buttonBaseClassName,
      color = '100',
      style,
      tag,
      subLabel,
      onClick,
      onCheckboxChange,
      checkboxProps,
      checkbox: hasCheckbox,
      indeterminate: isIndeterminate,
      ...props
    },
    ref
  ) => {
    const { classes, cx } = useStyles({ surface: color });

    const startIcon = 'startIcon' in props && props.startIcon;
    const endIcon = 'endIcon' in props && props.endIcon;
    const inlineMessage = 'inlineMessage' in props && props.inlineMessage;
    const buttonProps = 'buttonProps' in props && props.buttonProps;

    const renderLabel = () => {
      return typeof label === 'string' ? (
        <TruncatedText
          text={label}
          typographyProps={{
            component: 'span',
            weight: selected ? ('emphasis' as const) : ('regular' as const),
            className: labelClassName,
          }}
          tooltipProps={{ className: classes.label }}
        />
      ) : (
        label
      );
    };

    const renderTag = () => {
      return React.isValidElement(tag)
        ? React.cloneElement(tag, { size: 'sm', className: cx(tag.props.className, classes.tag) })
        : undefined;
    };

    const renderCheckboxLabel = () => {
      if (tag) {
        return (
          <span className={classes.checkboxWithTagWrapper}>
            {renderLabel()}
            {renderTag()}
          </span>
        );
      }
      return renderLabel();
    };

    const buttonBaseProps = useButtonBase(props);

    return (
      <MuiMenuItem
        ref={ref}
        classes={{ root: classes.root, selected: classes.selected, focusVisible: classes.focusVisible }}
        className={cx({ [classes.disabled]: disabled }, className)}
        selected={selected}
        disabled={disabled}
        component={elementType}
        style={style}
      >
        {hasCheckbox ? (
          <>
            {onCheckboxChange && (
              <Checkbox
                inputProps={checkboxProps}
                onChange={onCheckboxChange}
                className={classes.label}
                formControlLabelClassName={focusVisibleClassName}
                label={renderCheckboxLabel()}
                indeterminate={isIndeterminate}
                checked={selected}
                disabled={disabled}
              />
            )}
          </>
        ) : (
          <ButtonBase
            {...buttonProps}
            {...buttonBaseProps}
            color={color}
            onClick={onClick}
            className={cx(classes.wrapper, focusVisibleClassName, buttonBaseClassName)}
            focus="inset"
            disabled={disabled}
          >
            {startIcon && isValidElement(startIcon)
              ? React.cloneElement(startIcon, { size: 'md', className: classes.startIcon })
              : startIcon}
            <div className={cx(classes.labelWrapper, classes.label)}>
              {renderLabel()}
              {subLabel &&
                (typeof subLabel === 'string' ? (
                  <TruncatedText
                    text={subLabel}
                    typographyProps={{
                      component: 'div',
                      variant: 'body400',
                      secondary: true,
                      className: classes.subLabel,
                    }}
                  />
                ) : (
                  subLabel
                ))}
            </div>
            {inlineMessage &&
              React.isValidElement(inlineMessage) &&
              React.cloneElement(inlineMessage, {
                size: 'sm',
                className: cx(inlineMessage.props.className, classes.inlineMessage),
              })}
            {endIcon &&
              React.isValidElement(endIcon) &&
              React.cloneElement(endIcon, { size: 'md', className: cx(endIcon.props.className, classes.endIcon) })}
            {tag && renderTag()}
          </ButtonBase>
        )}
      </MuiMenuItem>
    );
  }
);
MenuItem.displayName = 'MenuItem';

export default React.memo(MenuItem);
