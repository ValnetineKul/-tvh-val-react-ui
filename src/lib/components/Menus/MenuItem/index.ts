import MenuItem from './MenuItem';

export type { MenuItemProps } from './MenuItem.types';
export default MenuItem;
