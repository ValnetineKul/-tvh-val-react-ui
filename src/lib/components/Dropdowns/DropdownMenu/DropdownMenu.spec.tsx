import React from 'react';
import { render, screen, userEvent, act, initResponsiveTest } from '../../../test-utils';

import DropdownMenu from './DropdownMenu';
import MenuItem from '../../Menus/MenuItem';

const props = {
  isOpen: true,
  menuItems: [<MenuItem label="item_1" onClick={jest.fn()} />, <MenuItem label="item_2" onClick={jest.fn()} />],
  onClose: jest.fn(),
};

describe('DropdownMenu', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  beforeAll(() => {
    jest.useFakeTimers();
  });
  afterAll(() => {
    jest.useRealTimers();
  });

  it('Should render menu items', () => {
    initResponsiveTest('Desktop');
    render(<DropdownMenu {...props} anchor={document.body} />);
    const dropdown = screen.getByRole('menu');
    expect(dropdown).toMatchSnapshot();
  });

  it('Should render sub menu', () => {
    render(
      <DropdownMenu
        {...props}
        anchor={document.body}
        menuItems={[
          ...props.menuItems,
          <MenuItem
            label="item_3"
            onClick={jest.fn()}
            subMenu={[<MenuItem label="sub_menu_1" onClick={jest.fn()} />]}
          />,
        ]}
      />
    );
    const subMenu = screen.getByText('item_3');
    act(() => {
      userEvent.click(subMenu);
    });
    expect(screen.getByText('sub_menu_1')).toBeInTheDocument();
  });

  it('Should call onClose when click outside', () => {
    initResponsiveTest('Desktop');
    const onClose = jest.fn();
    const anchor = document.createElement('div');
    render(<DropdownMenu {...props} onClose={onClose} anchor={anchor} />);
    jest.advanceTimersByTime(0);
    userEvent.click(document.body);
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('Should render Spinner if it is a loading state', () => {
    render(<DropdownMenu {...props} anchor={document.body} isLoading />);
    const spinner = screen.getByRole('progressbar');
    expect(spinner).toBeInTheDocument();
  });

  it('Should render Spinner if no menu items exist and it is a loading state', () => {
    render(<DropdownMenu {...props} anchor={document.body} isLoading menuItems={[]} />);
    const spinner = screen.getByRole('progressbar');
    expect(spinner).toBeInTheDocument();
  });

  it("Should not render Spinner when it isn't a loading state", () => {
    render(<DropdownMenu {...props} anchor={document.body} isLoading={false} />);
    const spinner = screen.queryByRole('progressbar');
    expect(spinner).not.toBeInTheDocument();
  });

  it("Should not render Spinner when no menu items exist and when it isn't a loading state", () => {
    render(<DropdownMenu {...props} anchor={document.body} isLoading={false} menuItems={[]} />);
    const spinner = screen.queryByRole('progressbar');
    expect(spinner).not.toBeInTheDocument();
  });

  it('Should disable sub menu', () => {
    render(
      <DropdownMenu
        {...props}
        anchor={document.body}
        menuItems={[
          ...props.menuItems,
          <MenuItem
            label="subMenu_item"
            onClick={jest.fn()}
            disabled
            subMenu={[<MenuItem label="sub_item_1" onClick={jest.fn()} />]}
          />,
        ]}
      />
    );

    expect(screen.getByText('subMenu_item').closest('button')).toBeDisabled();
  });

  describe('Should set correct component to display menu items', () => {
    it('Should show modal on mobile', () => {
      initResponsiveTest('Mobile');

      render(<DropdownMenu {...props} anchor={document.body} />);
      expect(screen.getByRole('dialog')).toBeInTheDocument();
      expect(screen.queryByRole('tooltip')).not.toBeInTheDocument();
    });

    it('Should show modal with sub menu on mobile', () => {
      initResponsiveTest('Mobile');

      render(
        <DropdownMenu
          {...props}
          anchor={document.body}
          menuItems={[
            ...props.menuItems,
            <MenuItem
              label="item_3"
              onClick={jest.fn()}
              subMenu={[<MenuItem label="sub_menu_1" onClick={jest.fn()} />]}
            />,
          ]}
        />
      );
      const subMenu = screen.getByText('item_3');
      userEvent.click(subMenu);
      expect(screen.getByText('sub_menu_1')).toBeInTheDocument();
    });

    it('Should show tooltip on tablet', () => {
      initResponsiveTest('Tablet');

      render(<DropdownMenu {...props} anchor={document.body} />);
      expect(screen.getByRole('tooltip')).toBeInTheDocument();
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });

    it('Should show tooltip on desktop', () => {
      initResponsiveTest('Desktop');

      render(<DropdownMenu {...props} anchor={document.body} />);
      expect(screen.getByRole('tooltip')).toBeInTheDocument();
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });
});
