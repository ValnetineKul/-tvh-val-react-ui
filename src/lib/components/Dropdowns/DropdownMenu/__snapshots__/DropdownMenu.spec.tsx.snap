// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu Should render menu items 1`] = `
<ul
  class="MuiList-root MuiList-padding tss-1j2rezf-DropdownMenu-menu-DropdownMenu-menuList-DropdownMenu-dynamicList mui-h4y409-MuiList-root"
  role="menu"
  tabindex="-1"
>
  <li
    class="tss-1fx4f83-MenuItem-root MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root  mui-t7a2rv-MuiButtonBase-root-MuiMenuItem-root"
    role="menuitem"
    tabindex="-1"
  >
    <button
      class="MuiButtonBase-root tss-7-MenuItem-wrapper-ref tss-19jpft6-ButtonBase-root-ButtonBase-focusInset-MenuItem-wrapper mui-10d1a0h-MuiButtonBase-root"
      tabindex="0"
      type="button"
    >
      <div
        class="tss-sqs6xo-MenuItem-labelWrapper-MenuItem-label"
      >
        <div
          class="tss-1-Typography-weightRegular-ref tss-m39tug-Typography-root-Typography-weightRegular-Tooltip-wrapper-TruncatedText-truncatedText-MenuItem-label-Typography-body400"
          data-mui-internal-clone-element="true"
        >
          <span
            class="tss-1-Typography-weightRegular-ref tss-b5vov3-Typography-root-Typography-weightRegular-TruncatedText-truncatedText-Typography-body500"
          >
            item_1
          </span>
        </div>
      </div>
    </button>
  </li>
  <li
    class="tss-1fx4f83-MenuItem-root MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root  mui-t7a2rv-MuiButtonBase-root-MuiMenuItem-root"
    role="menuitem"
    tabindex="-1"
  >
    <button
      class="MuiButtonBase-root tss-7-MenuItem-wrapper-ref tss-19jpft6-ButtonBase-root-ButtonBase-focusInset-MenuItem-wrapper mui-10d1a0h-MuiButtonBase-root"
      tabindex="0"
      type="button"
    >
      <div
        class="tss-sqs6xo-MenuItem-labelWrapper-MenuItem-label"
      >
        <div
          class="tss-1-Typography-weightRegular-ref tss-m39tug-Typography-root-Typography-weightRegular-Tooltip-wrapper-TruncatedText-truncatedText-MenuItem-label-Typography-body400"
          data-mui-internal-clone-element="true"
        >
          <span
            class="tss-1-Typography-weightRegular-ref tss-b5vov3-Typography-root-Typography-weightRegular-TruncatedText-truncatedText-Typography-body500"
          >
            item_2
          </span>
        </div>
      </div>
    </button>
  </li>
</ul>
`;
